docs/proposals/20240930-machine-drain-rules.md

```diff
@@ -0,0 +1,324 @@
+  * if the Pod has `cluster.x-k8s.io/drain-order: <order>`
+    * \=\> use `behavior: Drain` and `order: <order>`
+  * If there is a matching `MachineDrainRule`
+    * \=\> use `behavior` and `order` from the first matching `MachineDrainRule`
+  * Otherwise:
+    * \=\> use `behavior: Drain` and `order: 0`
+* If there are no more Pods to be drained
+  * \=\> drain is completed
+* Else:
+  * \=\> evict all Pods with the lowest `order`, update condition and requeue
+
+Notes:
+
+* It does not make sense to evict DaemonSet Pods, because the DaemonSet controller always [adds tolerations for the 
+  unschedulable taint to all its Pods](https://kubernetes.io/docs/concepts/workloads/controllers/daemonset/#taints-and-tolerations)
+  so they would be just immediately rescheduled.
+* It does not make sense to evict static Pods, because the kubelet won’t actually terminate them and the Pod object 
+  will get recreated.
+* If multiple `MachineDrainRules` match the same Pod, the “first” rule (based on alphabetical order) will be applied
+* `Pods` with drain behavior `Drain` and given order will be evicted only after all other `Pods` with a lower order 
+  have been already deleted
+* When waiting for volume detach, volumes belonging to `Pods` with drain behavior `Skip` are going to be ignored.
+  Or in other words, Machine deletion can continue if only volumes belonging to `Pods` with drain behavior `Skip`
+  are still attached to the Node.
+
+#### Changes to wait for volume detach
+
+Today, after Node drain we are waiting for **all** volumes to be detached. We are going to change that behavior to ignore
+all attached volumes that belong to Pods for which we skipped the drain.
+
+### Security Model
+
+This proposal will add a new `MachineDrainRule` CRD. The Cluster API core controller needs permissions to read this CRD.
+While permissions to write the CRD would be typically given to the cluster admin (the persona who has permissions to manage
+Cluster CRs), it is also possible to further restrict access.
+
+### Risks and Mitigations
+
+* There is ongoing work in Kubernetes in this area that we should eventually align with ([KEP-4212](https://github.com/kubernetes/enhancements/issues/4212),
+  [KEP-4563](https://github.com/kubernetes/enhancements/issues/4563))
+  * We are going to provide feedback to the KEPs to try to make sure the same use cases as in the current proposal are covered.
+
+## Alternatives
+
+**Add Node drain configuration to the Machine object**
+
+We considered adding the drain rules directly to the Machine objects instead. We discarded this option because it 
+would have made it necessary to configure Node drain on every single Machine. By having a separate CRD it is now 
+possible to configure the Node drain for all Clusters / Machines or a specific subset of Machines at once. This 
+also means that the Node drain configuration can be immediately changed without having to propagate configuration 
+changes to all Machines.
+
```

Comment:
I've another idea that we might want to note as a discarded alternative. Tracking the alternatives we discard helps (IMO) with the review and eventual approval.

- Define a CRD for MachineDrainRequest (_not_ NodeDrainRequest; this would be specific to Cluster API)
- Define a CRD for DrainRequestTemplate. This is the thing where you define how you want the Machine controller to drain Nodes. However, once you [as a controller :wink:] start a drain, you have a MachineDrainRequest and any changes at that stage would go into the MachineDrainRequest.

There can be rules within a MachineDrainRequest. Remember this is an alternative I propose that we discard, so it's not fully fleshed out. The MachineDrainRequest would fine all the drain request templates that look applicable and amalgamate those into one object, embedding the rules in those templates.

Downside: if you have lots of templates, you make a really large custom resource that is unwieldy to manage.

Reply:
I'll give it a try based on my understanding 😀